atlasdb-autobatch/src/main/java/com/palantir/atlasdb/autobatch/Autobatchers.java

```diff
@@ -126,19 +205,63 @@ private AutobatcherBuilder(Function<Integer, EventHandler<BatchElement<I, O>>> h
             return this;
         }
 
+        public AutobatcherBuilder<I, O> batchFunctionTimeout(Duration duration) {
+            this.batchFunctionTimeout = Optional.of(duration);
+            return this;
+        }
+
         public DisruptorAutobatcher<I, O> build() {
             Preconditions.checkArgument(purpose != null, "purpose must be provided");
 
-            int bufferSizeValue = bufferSize.orElse(DEFAULT_BUFFER_SIZE);
+            ImmutableEventHandlerParameters.Builder parametersBuilder = ImmutableEventHandlerParameters.builder();
+            bufferSize.ifPresent(parametersBuilder::batchSize);
+            parametersBuilder.safeLoggablePurpose(purpose);
+            Optional<TimeoutOrchestrationContext> timeoutOrchestrationContext =
+                    batchFunctionTimeout.map(timeout -> ImmutableTimeoutOrchestrationContext.builder()
+                            .batchFunctionTimeout(timeout)
+                            .exclusiveExecutor(PTExecutors.newCachedThreadPool("autobatcher." + purpose + "-timeout"))
```

Comment:
Yep, that's right. I'll answer 2 first as that's easier

> If we are not into sharing, why don't we use a single threaded executor -- aren't we guaranteed to only ever have one request in progress per autobatcher?

Strictly speaking, this is not necessarily true after this PR: if the batch function is _uninterruptible_, it is possible for us to have multiple requests in progress per autobatcher. (I've had a look in the Dialogue client and blocking clients are interruptible). In practice for most of our functions I'd expect one thread. What do you think?

> Why don't we share one across all autobatchers?

This allows better executor metrics (e.g. the safe loggable purpose) to be available in the event of a runaway executor. Other than that I didn't really have a reason

Reply:
(also, discussed offline: the current method actually shares them underneath the hood in terms of execution, because they use the Atlas view executor.)